q2_diversity_lib/tests/test_alpha.py

```diff
@@ -32,56 +28,70 @@ def setUp(self):
 
     def test_non_phylogenetic_passed_empty_table(self):
         for measure in nonphylogenetic_measures:
-            with self.assertRaisesRegex(ValueError, "empty"):
+            with self.assertRaisesRegex(ValueError, 'empty'):
                 measure(table=self.empty_table)
 
 
 class FaithPDTests(TestPluginBase):
-
     package = 'q2_diversity_lib.tests'
 
     def setUp(self):
         super().setUp()
-        self.input_table = biom.Table(np.array([[1, 0, .5, 999, 1],
-                                                [0, 1, 2, 0, 1],
-                                                [0, 0, 0, 1, 1]]),
-                                      ['A', 'B', 'C'],
-                                      ['S1', 'S2', 'S3', 'S4', 'S5'])
-        self.input_tree = skbio.TreeNode.read(io.StringIO(
-                '((A:0.3, B:0.50):0.2, C:100)root;'))
-        self.faith_pd_expected = pd.Series({'S1': 0.5, 'S2': 0.7, 'S3': 1.0,
-                                            'S4': 100.5, 'S5': 101},
-                                           name='faith_pd')
+        self.input_table_fp = self.get_data_path('faith_test_table.biom')
+        self.input_tree_fp = self.get_data_path('faith_test.tree')
+        self.rf_table = self.get_data_path('faith_test_table_rf.biom')
+        self.pa_table = self.get_data_path('faith_test_table_pa.biom')
+        self.empty_tree_fp = self.get_data_path('empty.tree')
+        self.root_only_tree_fp = self.get_data_path('root_only.tree')
+        self.missing_tip_tree_fp = self.get_data_path('missing_tip.tree')
+        self.expected = pd.Series({'S1': 0.5, 'S2': 0.7, 'S3': 1.0,
+                                   'S4': 100.5, 'S5': 101},
+                                  name='faith_pd')
 
     def test_receives_empty_table(self):
-        empty_table = biom.Table(np.array([]), [], [])
-        with self.assertRaisesRegex(ValueError, "empty"):
-            faith_pd(table=empty_table, phylogeny=self.input_tree)
+        # empty table generated from self.empty_table with biom v2.1.7
+        empty_table = self.get_data_path('empty_table.biom')
+        with self.assertRaisesRegex(ValueError, 'empty'):
+            faith_pd(table=empty_table, phylogeny=self.input_tree_fp)
```

Comment:
The view type of the table passed in here is a `str`, but the _registered_ view type for `faith_pd`'s `table` is `BIOMV210Format` - you will save yourself some time and heartache if you keep your tests consistent with the view declared in the signature (because the framework _will_ not allow this kind of sideways view coercion, so when you start wiring these functions up to pipelines you might see new errors, etc).

Reply:
and this probably means manually transforming my `str` into a BIOMV210Format with `transformFormat()`, or similar?

Reply:
Probably not - transformers are primarily invoked on `Artifact` instances --- for the sake of testing we usually just create a new instance of the format:

```python
empty_table_fp = self.get_data_path('empty_table.biom')
empty_table = BIOMV210Format(empty_table_fp, mode='r')
```

---

Alternatively, instead of testing the bare view function (which is what you're doing now `from q2_diversity_lib import faith_pd`) you could test the registered QIIME 2 action (see q2-cutadapt for an example: https://github.com/qiime2/q2-cutadapt/blob/master/q2_cutadapt/tests/test_demux.py). Then, instead of testing filepaths, you would actually pass your test data in in the form of a loaded Artifact.

The second approach has a lot of advantages - but most importantly it will save you from this case, where you have built tests that pass, but the tests themselves aren't actually representing the real-world case.

---

If you want a hands-on demonstration of this disconnect between your view annotation and the test suite/decorator logic, try running the following:

```bash
qiime diversity-lib faith-pd \
  --i-table table.qza \
  --i-phylogeny rooted-tree.qza \
  --o-vector faith.qza \
  --verbose
```

You'll see that the combination of registered view types, and the business logic in your decorators fails:

```bash
Traceback (most recent call last):
  File "/Users/matthew/.conda/envs/q2dev/lib/python3.6/site-packages/q2cli/commands.py", line 328, in __call__
    results = action(**arguments)
  File "<decorator-gen-491>", line 2, in faith_pd
  File "/Users/matthew/projects/qiime2/qiime2/qiime2/sdk/action.py", line 245, in bound_callable
    output_types, provenance)
  File "/Users/matthew/projects/qiime2/qiime2/qiime2/sdk/action.py", line 390, in _callable_executor_
    output_views = self._callable(**view_args)
  File "/Users/matthew/projects/qiime2/q2-diversity-lib/q2_diversity_lib/_util.py", line 63, in wrapper
    biom_table = load_table(table)
  File "/Users/matthew/.conda/envs/q2dev/lib/python3.6/site-packages/biom/parse.py", line 670, in load_table
    with biom_open(f) as fp:
  File "/Users/matthew/.conda/envs/q2dev/lib/python3.6/contextlib.py", line 81, in __enter__
    return next(self.gen)
  File "/Users/matthew/.conda/envs/q2dev/lib/python3.6/site-packages/biom/util.py", line 439, in biom_open
    if os.path.getsize(fp) == 0:
  File "/Users/matthew/.conda/envs/q2dev/lib/python3.6/genericpath.py", line 50, in getsize
    return os.stat(filename).st_size
TypeError: stat: path should be string, bytes, os.PathLike or integer, not BIOMV210Format

Plugin error from diversity-lib:

  stat: path should be string, bytes, os.PathLike or integer, not BIOMV210Format

See above for debug info.
```

You _won't_ see these failures in your test suite though, because you're testing something completely different (you're passing in strings of the filepath), instead of what you're telling QIIME 2 to do. Put another way, you have set up code that thinks it is looking at apples, and it advertises to the world that it should be given apples as input, but the business logic you have implemented, and the tests to test that business logic, are working with bananas. 🍌